fpsd/features.py

```diff
@@ -0,0 +1,765 @@
+
+        Args:
+            table_name [string]: Name of table to be dropped
+        """
+
+        query = "DROP TABLE IF EXISTS {} CASCADE;".format(table_name)
+        self.execute_query_from_string(query)
+        return None
+
+    def generate_table_undefended_frontpage_links(self):
+        """This method creates a table of exampleids that were
+        present at the time of feature generation that the unified
+        view can join on. The table contains one integer column, exampleid.
+        """
+
+        self.drop_stale_feature_table("undefended_frontpage_examples")
+
+        query = ("CREATE TABLE features.undefended_frontpage_examples AS ( "
+                 "SELECT foo.exampleid FROM ( SELECT exampleid, "
+                 "count(*) FROM raw.frontpage_traces GROUP BY exampleid) foo);")
+
+        self.execute_query_from_string(query)
+        return None
+
+    def _create_temp_packet_positions(self, outgoing_only=True):
+        """This method takes all rows in raw.frontpage_traces
+        and creates a temporary table packet_positions
+        with the following format:
+
+        exampleid  | rank     |  ingoing
+        (integer)  | (bigint) |  (boolean)
+        ------------------------------------
+        9          | 4        | f
+        9          | 5        | f
+        9          | 15       | f
+
+        where rank is the "position" in the trace for all
+        cells. For example, for the first outgoing cell in the
+        trace with exampleid=9 was the 4th packet in the trace.
+        This table is used by the position-based feature
+        generation functions.
+
+        Args:
+            outgoing_only [boolean]: describes whether or not we should
+                only use outgoing cells to create this table
+        """
+
+        self.drop_stale_feature_table("packet_positions")
+
+        if outgoing_only:
+            where_only_outgoing = "WHERE ingoing <= false"
+        else:
+            where_only_outgoing = ""
+
+        query = ("CREATE TEMP TABLE packet_positions AS    "
+                 "(SELECT exampleid, rank, ingoing                  "
+                 "  FROM (                                          "
+                 "    SELECT                                        "
+                 "      ROW_NUMBER() OVER                           "
+                 "      (PARTITION BY exampleid ORDER BY t_trace)   "
+                 "      AS rank,                                    "
+                 "      t.*                                         "
+                 "    FROM raw.frontpage_traces t)                  "
+                 "x {} );").format(where_only_outgoing)
+
+        self.execute_query_from_string(query)
+        return None
+
+    def _create_table_outgoing_cell_positions(self, num_cells=500):
+        """This method takes the first num_cells rows in
+        packet_positions and creates a temporary
+        table top_(num_cells)_outgoing_packet_positions
```

Comment:
* I like this suggestion, fixed

* True, fixed the docstring to explicitly say this

* We _could_, but building this as a view would make queries on this data slower. Also note temporary tables get auto dropped at the end of a session so there’s really not much advantage to building views on them (at least in this situation).